### PR TITLE
feat: CLIN-3406 make vep command configurable

### PR DESCRIPTION
**Pull request by LysianeBouchard**
The goal of this pull request is to allow the VEP process to be easily adapted for different projects (e.g., QLIN, CQDG). We should be able to use different VEP commands and versions for each project.

To achieve this, we avoid hardcoding any arguments that are not likely to be shared between projects in the process script. Instead, these arguments can be configured via the task.ext.args directive.

For now, to maintain the current behaviour, which is aligned to CQDG, we will include the arguments from the current VEP command (CQDG) by default in the configuration. This behaviour can be overridden in QLIN by providing an extra config file at the command line.

For the VEP version, one can simply overwrite the container directive in the process configuration. We will continue to use the same container as for CQDG (version 111) by default.   We might need to build our own containers to support older VEP versions, as I am not sure that plugins are installed by default for these versions. We will address this issue in a separate PR if needed.

**Other changes:**
- We took the opportunity to replace our local VEP module with a more standard nf-core module.
- After removing previous vep process, moved remaining processes in the VEP module to dedicated module files; it was odd having a module called VEP without a VEP process inside.
- Improve the parameter schema  for parameters max_memory, max_time and max_disk
- Made a small change to return the exomiser output channel in the main post-processing workflow.  It might be useful if we need to add extra steps in the main workflow
- Fix link pointing to incorrect vep version in vep reference data documentation

**Tests**

- **Run the pipeline locally using our public test dataset**:
`nextflow run main.nf -profile test,docker`
Check that we can see the vep output files in the results, with the .tbi files stored in the same folder. 
Check that the .tbi files are named consistently with corresponding vcf files.
Run a few bcftools view commands.

- **Compare results with results obtain from the main branch using diff commands**:
```
mv results results_new
git checkout main
nextflow run main.nf -profile test,docker

# compare vcf files body, i.e. without the header
bcftools view -H results/vep/variants.family1.vep.vcf.gz >family1_prev_body
bcftools view -H results_new/ensemblvep/variants.family1.vep.vcf.gz >family1_new_body
bcftools view -H results/vep/variants.family2.vep.vcf.gz >family2_prev_body
bcftools view -H results_new/ensemblvep/variants.family2.vep.vcf.gz >family2_new_body    
diff family1_prev_body family1_new_body    
diff family2_prev_body family2_new_body   

# compare vcf files header
bcftools view -h results/vep/variants.family1.vep.vcf.gz >family1_prev_header
bcftools view -h results_new/ensemblvep/variants.family1.vep.vcf.gz >family1_new_header
bcftools view -h results/vep/variants.family2.vep.vcf.gz >family2_prev_header
bcftools view -h results_new/ensemblvep/variants.family2.vep.vcf.gz >family2_new_header   
diff family1_prev_header family1_new_header
diff family2_prev_header family2_new_header 
```
The vcf bodies were identical. 
There are differences in headers, but they should cause no harm:
- dates differ
- in the combine, the header of one of the 3  input file is copied in the final trio file. It's not always the same input file that is picked, so the filenames in the stored commands change
- the vep command changes slightly, which is normal as the order of the arguments changed

- **Test changing the docker container / command**

I am able to change the container to v103 and use the same command.
I am able to change the container to v103 and change the command.

**Tests in juno**

Run the pipeline in juno, and compare the results with the main branch. I see the same behaviour observed locally, i.e. the vcf bodies are identical. There are harmless differences in headers. The corresponding .tbi files are published correctly within the same folder.

I also check that the pipeline validation logic was correctly applied for the vep parameters. When I run a command with missing parameter, it quickly fail with an error message indicating that required parameters are missing.

Replicated tests to change container and command.

**Extra linter warnings introduce**
This PR introduces new linter warnings.  I moved local processes originally in the vep.nf module file into dedicated module files. As these processes have not been standardized to nf-core yet, the associated non conformity warnings now appear once for each module. The new warnings are actually more representative of the true state of the code. They will be removed when we standardize these processes, as we plan to do.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/ferlab/postprocessing/tree/master/.github/CONTRIBUTING.md)
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] Reference Data Documentation in `docs/reference_data.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: conf/modules.config [hunk 1]
```diff
@@ -57,11 +57,32 @@ process {
...
+            "--hgvsg",
+            "--canonical",
+            "--symbol",
+            "--flag_pick",
+            "--no_stats",
+            "--fields \"Allele,Consequence,IMPACT,SYMBOL,Feature_type,Gene,PICK,Feature,EXON,BIOTYPE,INTRON,HGVSc,HGVSp,STRAND,CDS_position,cDNA_position,Protein_position,Amino_acids,Codons,VARIANT_CLASS,HGVSg,CANONICAL,RefSeq\""
+        ]
+        ext.args = args_list.join(" ")
+        ext.prefix =  {"variants.${meta.id}.vep"}
+    }
```
[the remark this answers is not in the document]
Comment (LysianeBouchard) [reply conf/modules.config:79]: Here is the content of the script block (nf-core vep) on this branch:
  ```
def args = task.ext.args ?: ''
    def file_extension = args.contains("--vcf") ? 'vcf' : args.contains("--json")? 'json' : args.contains("--tab")? 'tab' : 'vcf'
    def compress_cmd = args.contains("--compress_output") ? '' : '--compress_output bgzip'
    def prefix = task.ext.prefix ?: "${meta.id}"
    def dir_cache = cache ? "\${PWD}/${cache}" : "/.vep"
    def reference = fasta ? "--fasta $fasta" : ""
    """
    vep \\
        -i $vcf \\
        -o ${prefix}.${file_extension}.gz \\
        $args \\
        $compress_cmd \\
        $reference \\
        --assembly $genome \\
        --species $species \\
        --cache \\
        --cache_version $cache_version \\
        --dir_cache $dir_cache \\
        --fork $task.cpus

```

Comment (LysianeBouchard) [inline conf/modules.config:79]: These  arguments will be added to the vep command. To ease the comparison, I copy-pasted final vep commands in vcf headers obtained running the pipeline on the main branch and this branch. This way it is easier to review the final command.

On the main branch, we see this command (reformatted to put one option per line):
```
vep 
  --cache 
  --canonical 
  --compress_output bgzip 
  --database 0 
  --dir vep 
  --dir_plugins /plugins 
  --fasta [PATH]/Homo_sapiens_assembly38.fasta 
  --fields Allele,Consequence,IMPACT,SYMBOL,Feature_type,Gene,PICK,Feature,EXON,BIOTYPE,INTRON,HGVSc,HGVSp,STRAND,CDS_position,cDNA_position,Protein_position,Amino_acids,Codons,VARIANT_CLASS,HGVSg,CANONICAL,RefSeq 
  --flag_pick 
  --fork 4 
  --format vcf 
  --hgvs 
  --hgvsg 
  --input_file family2.splitted.vcf.gz 
  --no_stats 
  --numbers 
  --offline 
  --output_file variants.family2.vep.vcf.gz 
  --symbol 
  --variant_class 
  --vcf 
  --xref_refseq'
```

File: conf/modules.config [hunk 2]
```diff
@@ -57,11 +57,32 @@ process {
         ]
     }
 
+    withName: ENSEMBLVEP_VEP {
+        container = 'ensemblorg/ensembl-vep:release_111.0' //sticking to v111 for now, but we should update this
```
Comment (DamienGnst) [inline conf/modules.config:61]: This version should match the cache_version  or at least this one can't be inferior to the cache version, maybe we need to define the  container version in the module ..

File: workflows/postprocessing.nf [hunk 3]
```diff
@@ -11,9 +11,9 @@ include { softwareVersionsToYAML } from '../subworkflows/nf-core/utils_nfcore_pi
 include { EXCLUDE_MNPS           } from "../subworkflows/local/exclude_mnps"
 include { VQSR                   } from "../subworkflows/local/vqsr"
 include { EXOMISER               } from '../modules/local/exomiser'
-include { splitMultiAllelics     } from '../modules/local/vep'
-include { vep                    } from '../modules/local/vep'
-include { tabix                  } from '../modules/local/vep'
+include { splitMultiAllelics     } from '../modules/local/split_multi_allelics'
+include { ENSEMBLVEP_VEP } from '../modules/nf-core/ensemblvep/vep/main'
```
Comment (DamienGnst) [inline workflows/postprocessing.nf:15]: just an indent to make it cleaner

File: workflows/postprocessing.nf [hunk 4]
```diff
@@ -86,6 +86,25 @@ def exomiser(inputChannel,
...
+    def ch_input_for_vep  = input_channel.map{meta, files ->
+        def vcf_file = files.find { it.name.endsWith("vcf.gz") }
+        def custom_extra_files = [] 
+        [meta, vcf_file, custom_extra_files]
+    }
+  
+    return ENSEMBLVEP_VEP(
+        ch_input_for_vep, 
+        vep_genome,
+        "homo_sapiens", //species
```
Comment (DamienGnst) [inline workflows/postprocessing.nf:100]: should we use an arg to define the species.

File: conf/modules.config (hunk 1 above)
Comment (FelixAntoineLeSieur) [reply conf/modules.config:79]: I think those are some good changes I just don't understand why the fork is being set to 2 when it seems to be 4 in the nextflow.config file, under the ENSEMBL_VEP process definition?

File: workflows/postprocessing.nf (hunk 4 above)
Comment (FelixAntoineLeSieur) [reply workflows/postprocessing.nf:100]: maybe, but I feel it would be unlikely to be used on non-humans, no?

**Review by FelixAntoineLeSieur — APPROVED** [review]

File: conf/modules.config (hunk 2 above)
Comment (LysianeBouchard) [reply conf/modules.config:61]: We were using version 111 before. If I understand this comment properly, our current vep reference data in cqdg match version 112 and not 111. I will try to figure out which vep version is the reference data. There might be a metadata file / checksum that I can use.

Comment (LysianeBouchard) [reply conf/modules.config:61]: Pour l'instant, je vois que le nom du folder qui contient les données de vep dans le minio de CQDG est bien nommé selon la version 111:
cqdg-prod-file-import/nextflow/reference/annotation/vep/homo_sapiens/111_GRCh38/

Je suis en train de télécharger le fichier de vep pour comparer. La vitesse de téléchargement est très lente. Ça va probablement prendre toute la nuit.

File: conf/modules.config (hunk 1 above)
Comment (LysianeBouchard) [reply conf/modules.config:79]: Bonne question. Je crois que c'est ce paramètre dans le test config qui crée ça:  `max_cpus   = 2`

Voici ce qui est configuré dans nextflow.config pour le process ENSEMBL_VEP:
cpus	=	{ check_max( 4 		* task.attempt, 'cpus'		)	}

Dans la fonction check_max, la valeur 4 va être remplacée par  `Math.min( obj, params.max_cpus as int )`

File: workflows/postprocessing.nf (hunk 3 above)
Comment (LysianeBouchard) [reply workflows/postprocessing.nf:15]: will do

File: workflows/postprocessing.nf (hunk 4 above)
Comment (LysianeBouchard) [reply workflows/postprocessing.nf:100]: Oui c'est pour cette raison que j'avais hard codé le tout. Sans nécéssairement introduire un paramètre, je pourrais utiliser une constante ou juste une variable. Ça clarifierait qu'il s'agit d'une espèce. Je vais voir ce que je peux faire pour améliorer ici.

Comment (LysianeBouchard) [reply workflows/postprocessing.nf:100]: J'ai simplement créé une constante dans le fichier pour l'instant.

Comment (LysianeBouchard) [reply workflows/postprocessing.nf:100]: Hum ... après avoir testé, je vois que ça ne fonctionne pas avec la constante. Je ne sais pas trop pourquoi, mais elle ne semble pas accessible dans la fonction `vep`, sauf si je la mets globale (sans le mot clé def), ce que je n'aime pas trop.

Par contre, elle est accessible dans le corps du workflow.  Si je la passe en argument à la fonction vep à partir du workflow, là ça fonctionne.  C'est ce que j'ai fait pour l'instant. Je vais re-tester sur CQDG pour m'assurer que c'est ok dans un autre contexte d'exécution.

File: workflows/postprocessing.nf (hunk 3 above)
Comment (LysianeBouchard) [reply workflows/postprocessing.nf]: done

File: conf/modules.config (hunk 2 above)
Comment (LysianeBouchard) [reply conf/modules.config:61]: J'ai pu voir ce matin que la taille du vep cache qu'on a sur JUNO est plus grande (environ 26G) que celle de la version 111 (environ 23G). 

Je crois qu'il s'agit probablement de la version 112.  Je vais corriger la documentation pour que ce soit clair et je vais ouvrir un ticket séparé pour fixer ça.

**Comment by LysianeBouchard** [comment]
J'ai ajouté un changement supplémentaire dans le pull request pour ne considérer que les versions stables de Nextflow dans les tests GitHub CI. Auparavant, nous considérions également les versions edge, ce qui, à mon avis, n'est pas très prudent. Nous utilisons notamment une action de nf-core pour installer Nextflow. S'il y a des changements breaking dans une version edge, il est fort probable qu'ils n'aient pas encore eu le temps de faire une mise à jour.

**Comment by dmorais** [comment]
The code improves the pipeline flow and allow for a more configurable VEP versioning

**Review by dmorais — APPROVED** [review]
The code improves pipeline readability and flexibility in terms of VEP versioning.

**Pull request merged**